docs/proposals/custom-metrics.md

```diff
@@ -0,0 +1,119 @@
+Custom metrics are needed by:
+* horizontal pod autoscaler, for autoscaling the number of pods based on them;
+* scheduler, for using them in a sophisticated scheduling algorithm;
+* general monitoring of application health by a user (we are not sure yet whether this should be handled separately, by a different,
+general monitoring solution, see [#665](https://github.com/kubernetes/heapster/issues/665));
+* ...
+
+
+## Design
+
+For the Kubernetes version 1.2, we plan to implement aggregation of pod custom metrics in Prometheus format by pull.
+
+Each pod, to expose custom metrics, will expose a set of Prometheus endpoints.
+CAdvisor will collect metrics from such endpoints of all pods on each node by pulling, and expose them to Heapster.
+Heapster will:
+* collect custom metrics from all CAdvisors in the cluster, together with pulling system metrics
+(for version 1.2: we assume pooling period of ~30 seconds and the number of custom metrics ~3 per pod,
+we are not going to enforce the number by any quota),
+* store them in a metrics backend (influxDB, Prometheus, Hawkular, GCM, …),
+* expose the latest snapshot of custom metrics for queries (by HPA/scheduler/…) using model API.
+
+User can easily expose Prometheus metrics for her own application by using Prometheus [client](http://prometheus.io/docs/instrumenting/clientlibs/) library.
+To monitor third-party applications, Prometheus [exporters](http://prometheus.io/docs/instrumenting/exporters/) run as side-cars containers may be used.
+
+In future versions (later than 1.2), we want to extend our solution by:
+* accepting pod metrics exposed in different formats than Prometheus
+(collecting of the different formats will need to be supported by cAdvisor),
+* support push metrics by exposing push API on heapster (e.g. in StatsD format) or on a local node collector
+(if heapster performance is insufficient),
+* support standalone custom metrics (not connected to a pod).
+
+
+## API
+
+For Kubernetes 1.2, defining pod Prometeus endpoints we be done using annotations.
+Later, when we are sure that our API is correct and stable, we will make it a part of `PodSpec`.
+
+We will add a new optional pod annotation with the following key: `metrics.alpha.kubernetes.io/custom-endpoints`.
+It will contain a string-value in JSON format.
+The value will be a list of tuples defining ports, paths and protocol
+(currently, we will support only Prometheus protocol, this will be also the default value if format is empty)
+of metrics endpoints exposed by the pod, and names of metrics which should be taken
+(if empty: we will take all exposed metrics) from the endpoint.
+
+For example, the following configuration:
+
+```
+"metrics.alpha.kubernetes.io/custom-endpoints" = [
+	{
+		"protocol": "prometheus",
+		"path": "/status",
+		"port": "8080",
+		"names:" ["qps", "activeConnections"]
+	},
+	{
+		"path": "/metrics",
+		"port": "9090",
+	}
+]
```

Comment:
I'd argue that have embedded JSON that needs to be written by humans is not the best idea (having to escape quotes by hand and all that), but the alternatives are not nearly as nice.